### ISSUE TITLE
Better handling of MultipleRunningTimestampServiceError

### ISSUE THREAD

**Issue by stefank-palantir** [palantir/atlasdb#1749]
@jboreiko identified an issue we think caused PDS-50340, whereby a leadership election occurs whilst the immutable timestamp is being updated by the existing leader, and a race condition causes a runtime exception in the existing leader. This causes the client call to fail at the very least, and possibly (?) death of the server (John and I weren't clear on the worst-case impact).

Since clustered lock should be highly available, this feels like a bad bug.

@jboreiko , if you could flesh out this explanation it would be great – but I believe the full timeline (assuming a SQL database) is:
* Server A is leader and is running low on timestamps
* Server A receives a request that requires the immutable timestamp to be updated
* Server A makes a SQL call to the underlying DB to update the timestamp bound
* A leadership election occurs, for whatever reason
* Server B is elected leader
* Server B tries to grab initial timestamps
* Server B makes a call to the DB to update the timestamp bound
* Server B's call completes
* Server A's DB transaction is now conflicted and fails
* This triggers a `MultipleRunningTimestampServiceError` in A

The change we discussed briefly was to make `MultipleRunningTimestampServiceError` a subclass of `NotCurrentLeaderException`. We think this should cause the client to failover cleanly to the new leader. However if for some reason there _are_ two running leaders this could worsen the impact, so we should think carefully about the minimum fix.

From a Support perspective this is a P1 issue for the customer due to having hard availability requirements, so it would be good to prioritize this if we can.

**Comment by hsaraogi** [comment]
This seems quite unlikely with timelock server, but we want to track it to dig into if observed in field again.

**Comment by ryanfantus** [comment]
@hsaraogi may have hit this just now - https://support-jira.yojoe.local/browse/PDS-51875

**Comment by tpetracca** [comment]
Worth noting the case @ryanfantus is referencing was using external TimeLock

**Comment by jeremyk-91** [comment]
This issue looks like it occurred again on a stack with external TimeLock (internal reference PDS-52229; consult the ticket for more detailed discussion)

**Comment by gsheasby** [comment]
In that issue, it was the scenario detailed in the ticket description.

**Proposal: When we fail to update the bound, log a warning, but still hand out timestamps we have already reserved**

The scenario above would then look like:
- Server A has reserved timestamps up to 1337 (bound is "A, 1337"), and has handed out timestamps up to 1086.
- Server A receives a request for 3 timestamps. This requires the timestamp bound to be updated
- Server A makes a proposal to update the timestamp bound (to reserve up to 1339)
- A leadership election occurs, for whatever reason
- Server B is elected leader
- Server B tries to grab initial timestamps (1338 and 1339)
- Server B makes a call to the DB to update the timestamp bound (to 1339)
- Server B's call completes. Bound is now "B, 1339"
- Server A's proposal fails, because "B, 1339" has been accepted
- Instead of throwing MRTSE, Server A logs a message, and fulfils the requests, giving timestamps 1087 to 1089
- Other in-flight requests to Server A might get timestamps between 1090 and 1337, but **will not get timestamps greater than 1337** (and no more requests will be added to the queue).
- Meanwhile, B will serve requests, and give out timestamps from 1340+

Two concurrently running threads, one talking to A and one talking to B, may get non-sequential timestamps, but each thread's individual view will have increasing timestamps.

Our transactional guarantee is that each _thread_ sees monotonically increasing timestamps, and each timestamp handed out is unique. Note that we can't make the global guarantee, even if both threads are talking to the same server.

"if for some reason there are two running leaders this could worsen the impact" - thanks to the Paxos guarantees (and given sane config), there can't be two running leaders simultaneously, as any node requires the agreement of a quorum of nodes to become the leader.

There are pathological cases where bad config and bad luck can make this guarantee fail, but protecting against that is beyond the scope of this ticket.

**Comment by nziebart** [comment]
I think @gsheasby's proposal makes sense. As long as Server A was still the leader at the start of the request, it should be able to return timestamps it has already allocated.

However, I think it's also possible for the leader to hit this error (before it has allocated any timestamps): after an election, the old leader may still be servicing a request, and may perform a CAS before the new leader does.

As a solution, we could verify leadership after hitting a `MultipleRunningTimestampServiceError`. If we hit this error, and are no longer the leader, we can return a 503. If we are the leader, we can retry the CAS, assuming that some other non-leader node has pre-empted us. 

I believe the [verify leadership, attempt CAS, verify leadership] sequence serves as a fencing mechanism, and provides the guarantee that we will have to retry at most once, otherwise there is actually another rogue process modifying the timestamp bound.

**Comment by gsheasby** [comment]
>  I think it's also possible for the leader to hit this error (before it has allocated any timestamps): after an election, the old leader may still be servicing a request, and may perform a CAS before the new leader does.

That case would be:
* Server A has reserved timestamps up to 1337 (bound is "A, 1337"), and has handed out timestamps up to 1086.
* Server A receives a request for 3 timestamps. This requires the timestamp bound to be updated
* Server A makes a proposal to update the timestamp bound (to reserve up to 1339)
A leadership election occurs, for whatever reason
* Server B is elected leader
* Server B gets the latest bound and finds it to be "A, 1337"
* Server B makes a call to the DB to update the timestamp bound (to 1339)
* Server A's call completes. Bound is now "A, 1339"
* Server B's proposal fails, because "A, 1339" has been accepted

In this case, I think that Server B should attempt to recover. We might also want to invalidate the previous update (A, 1339), because otherwise A might hand out timestamps 1338 and 1339, which were already grabbed by B.

This means that the following for A:
* Server A "remembers" a bound of (A, 1337)
* Server A proposes to update the bound to (A, 1339)
* Server A's call completes (bound is now "A, 1339")
* Server A verifies that it is still leader, and finds that it is not.
* Server A does not store the remembered bound in memory - its view of the bound remains 1337. Further, Server A does not attempt to update the bound again.

And for B:
* Server B is elected leader
* (1) Server B gets the latest bound and finds it to be "A, 1337"
* Server B makes a call to the DB to update the timestamp bound (to 1339)
* Server A's call completes. Bound is now "A, 1339"
* Server B's proposal fails, because "A, 1339" has been accepted
* Server B verifies that it is still leader, and finds that it *is*.
* Server B goes back to 1, except now it will find the bound to be (A, 1339), and will propose (B, 1341).

**Comment by nziebart** [comment]
It looks like the linked PR fixed this for the case where the old leader fails to CAS, but doesn't fix the case where the new leader fails (i.e., the second scenario described above).

We may want to keep this open until we have tests for these scenarios and are confident that we're handling all valid cases.

**Comment by nziebart** [comment]
Talked with @fsamuel-bs offline, he mentioned that the reasoning here was that, even if the new leader fails on CAS (because the old leader succeeded), it can throw a 503 which will cause the request to get retried. 

However, I don't think actually works because we will attempt to CAS from the current value, and that value is only updated (1) when the timestamp service is first created and (2) when we perform a successful CAS. Thus, we'll just throw 503's indefinitely. The test below goes into an infinite loop:
```
    public void testRetryStoreUpperLimit() throws InterruptedException {
        PersistentTimestampService oldLeader = PersistentTimestampService.create(createPaxosTimestampBoundStore(0));
        PersistentTimestampService newLeader = PersistentTimestampService.create(createPaxosTimestampBoundStore(1));

        // in-progress request on the old leader
        oldLeader.getFreshTimestamp();

        // request to the new leader
        while(true) {
            try {
                newLeader.getFreshTimestamp();
            } catch (Throwable ex) {
                if (ex.getCause() instanceof NotCurrentLeaderException) {
                    System.out.println("not the leader; retrying");
                } else {
                    throw new RuntimeException(ex);
                }
            }
        }
    }
```

As a side note, the `NotCurrentLeaderException` is being wrapped in a `RuntimeException` by `AllocationFailures#responseTo` - so I'm not sure if this actually gets correctly propagated as a 503

**Comment by nziebart** [comment]
Also, it looks like we were already mapping `MultipleRunningTimestampServiceError` to a 503 in `TimestampAllocationFailures`, before this issue was created.

**Comment by nziebart** [comment]
I see now that `AwaitingLeadershipProxy` will re-create the delegate if you throw a `ServiceUnavailableException` or `NotCurrentLeaderException` - so the proposed fix here should handle both cases. So, given that we were apparently already doing this, the referenced ticket may have been caused by something else.

**Comment by jeremyk-91** [comment]
> As a side note, the NotCurrentLeaderException is being wrapped in a RuntimeException by AllocationFailures#responseTo - so I'm not sure if this actually gets correctly propagated as a 503

Right - I subsequently tested this and you would get a 500 in this case. (This is actually OK in terms of what a client would do, for the most part, but we're relying on clients retrying sensibly, which is not great.) Fixed in 0.42.2 and will send this to develop tomorrow as well.

> Also, it looks like we were already mapping MultipleRunningTimestampServiceError to a 503 in TimestampAllocationFailures, before this issue was created.

We were, as you've pointed out. I think there's a bit of (probably awful and unwanted) complexity here though; a "fast failover" exception where you immediately go and talk to other nodes is one with a Retry-After header (presumably of zero, though we don't actually check that yet). Previous MRTSEs would have been a 503 with no Retry-After header.
